applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/formfinding_strategy.hpp

```diff
@@ -0,0 +1,265 @@
+                    CalculateReactions,
+                    ReformDofSetAtEachStep,
+                    MoveMeshFlag),
+                    mProjectionSettings(ProjectionSetting),
+                    mrFormFindingModelPart(rFormFindingModelPart),
+                    mPrintingFormat(PrintingFormat),
+                    mWriteNewReferenceFile(WriteNewReferenceFile)
+                 {
+                    InitializeIterationIO();
+                 }
+
+            // constructor with Builder and Solver
+            FormfindingStrategy(
+                ModelPart& model_part,
+                typename TSchemeType::Pointer pScheme,
+                typename TLinearSolver::Pointer pNewLinearSolver,
+                typename TConvergenceCriteriaType::Pointer pNewConvergenceCriteria,
+                typename TBuilderAndSolverType::Pointer pNewBuilderAndSolver,
+                ModelPart& rFormFindingModelPart,
+                bool WriteNewReferenceFile,
+                std::string PrintingFormat,
+                Parameters ProjectionSetting,
+                int MaxIterations = 30,
+                bool CalculateReactions = false,
+                bool ReformDofSetAtEachStep = false,
+                bool MoveMeshFlag = false
+                )
+                : ResidualBasedNewtonRaphsonStrategy<TSparseSpace, TDenseSpace, TLinearSolver>(model_part, pScheme,
+                    pNewLinearSolver,pNewConvergenceCriteria,pNewBuilderAndSolver,MaxIterations,CalculateReactions,ReformDofSetAtEachStep,
+                    MoveMeshFlag),
+                    mProjectionSettings(ProjectionSetting),
+                    mrFormFindingModelPart(rFormFindingModelPart),
+                    mPrintingFormat(PrintingFormat),
+                    mWriteNewReferenceFile(WriteNewReferenceFile)
+                 {
+                    InitializeIterationIO();
+                 }
+
+            // Destructor
+            ~FormfindingStrategy() = default;
+
+        private:
+            bool SolveSolutionStep() override
+            {
+                BaseType::SolveSolutionStep();
+                if (mPrintingFormat=="all" || mPrintingFormat=="gid") mpIterationIO->FinalizeResults();
+                return true;
+            }
+
+            void UpdateDatabase(
+                    TSystemMatrixType& A,
+                    TSystemVectorType& Dx,
+                    TSystemVectorType& b,
+                    const bool MoveMesh
+                ) override
+                {
+                    BaseType::UpdateDatabase(A,Dx, b, MoveMesh);
+                    for(auto& r_node : mrFormFindingModelPart.Nodes()){
+                        // Updating reference
+                        const array_1d<double, 3>& disp = r_node.FastGetSolutionStepValue(DISPLACEMENT);
+                        array_1d<double, 3>& disp_non_historical = r_node.GetValue(DISPLACEMENT);
+
+                        disp_non_historical = disp_non_historical + disp;
```

Comment:
I basically like the idea that you exchange the historical variable each time but keep one for the comparison to initial.

However, this will be definitely a problem when it comes to custom io's. I would maybe suggest introducing another dof (should be the same memory) e.g. DISPLACEMENT_TO_INITIAL_CONFIGURATION. (Maybe you'll find a better name).

That would avoid a lot of confusion and would allow that you basically don't have to keep consistency to the io's.

Reply:
yes, we thought about that. But I think using the non-historical variable is the better choice. We do not need a new variable and it is only used to visualize the form-finding process. You actually don't need it for anything else.

Reply:
If you embed this in a historical simulation (as IGA is intrinsically) you require this and as mentioned if something changes within the io's, you do not consider formfinding specifically. Especially, as this wouldn't be used all times. Introducing anther variable is very simple and would simplify a lot.

Reply:
yeah, makes sense. How about DISPLACEMENT_FORMFINDING

Reply:
@tteschemacher I don't really follow
I suggest to change this when it becomes necessary, it sounds like there will be some discussions necessary
Since it is only used for visualization atm it can be refactored easily